File: solutions/2501. Longest Square Streak in an Array/2501.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  int longestSquareStreak(const vector<int>& nums) {
    int ans = 0;
    unordered_set<int> uniqueNums(nums.begin(), nums.end());

    for (const int num : nums) {
      int currStreak = 0;
      long curr = num;

      while (uniqueNums.find((int)curr) != uniqueNums.end()) {
        ++currStreak;
        if (curr * curr > 100'000)
          break;
        curr *= curr;
      }
      ans = max(ans, currStreak);
    }

    return ans < 2 ? -1 : ans;
  }
};

File: solutions/2501. Longest Square Streak in an Array/2501.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def longestSquareStreak(self, nums: List[int]) -> int:
    ans = 0
    uniqueNums = set(nums)

    for num in nums:
      currStreak = 0
      curr = num

      while curr in uniqueNums:
        currStreak += 1
        if curr * curr > 100_000:
          break
        curr *= curr

      ans = max(ans, currStreak)

    return ans if ans >= 2 else -1